File: core/tests/helpers/signal/signal_guard_test.c
/**
 * Pure C test for signal guard SIGBUS recovery.
 *
 * Run via CI script:
 *   ./scripts/check_signal_tests.sh
 *
 * Or manually:
 *   gcc -o signal_guard_test_c core/tests/helpers/signal/signal_guard_test.c core/src/capi/signal_guard.c
 *   ./signal_guard_test_c
 */

#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>

/* Import from signal_guard.c */
extern void talu_signal_guard_install(void);
extern int talu_signal_guard_call(int (*func)(void*), void* ctx);

/* Test callbacks */
static int normal_callback(void* ctx) {
    (void)ctx;
    /* Just return success */
    return 0;
}

static int sigbus_callback(void* ctx) {
    (void)ctx;
    /* Raise SIGBUS - should be caught */
    raise(SIGBUS);
    /* Should NOT reach here */
    return 0;
}

static int returns_42(void* ctx) {
    (void)ctx;
    return 42;
}

int main(void) {
    talu_signal_guard_install();

    /* Test 1: Normal operation */
    printf("Test 1: Normal operation... ");
    fflush(stdout);
    {
        int result = talu_signal_guard_call(normal_callback, NULL);
        if (result == 0) {
            printf("PASS\n");
        } else {
            printf("FAIL (expected 0, got %d)\n", result);
            return 1;
        }
    }

    /* Test 2: SIGBUS recovery */
    printf("Test 2: SIGBUS recovery... ");
    fflush(stdout);
    {
        int result = talu_signal_guard_call(sigbus_callback, NULL);
        if (result == -1) {
            printf("PASS\n");
        } else {
            printf("FAIL (expected -1, got %d)\n", result);
            return 1;
        }
    }

    /* Test 3: Multiple cycles */
    printf("Test 3: Multiple cycles... ");
    fflush(stdout);
    {
        /* Cycle 1 - normal */
        int result = talu_signal_guard_call(normal_callback, NULL);
        if (result != 0) {
            printf("FAIL (cycle 1: expected 0, got %d)\n", result);
            return 1;
        }

        /* Cycle 2 - with signal */
        result = talu_signal_guard_call(sigbus_callback, NULL);
        if (result != -1) {
            printf("FAIL (cycle 2: expected -1, got %d)\n", result);
            return 1;
        }

        /* Cycle 3 - normal again */
        result = talu_signal_guard_call(normal_callback, NULL);
        if (result != 0) {
            printf("FAIL (cycle 3: expected 0, got %d)\n", result);
            return 1;
        }

        printf("PASS\n");
    }

    /* Test 4: Return value propagation */
    printf("Test 4: Return value propagation... ");
    fflush(stdout);
    {
        int result = talu_signal_guard_call(returns_42, NULL);
        if (result == 42) {
            printf("PASS\n");
        } else {
            printf("FAIL (expected 42, got %d)\n", result);
            return 1;
        }
    }

    printf("\nAll tests passed!\n");
    return 0;
}

File: core/tests/helpers/signal/signal_guard_test.zig
//! Standalone test for SignalGuard SIGBUS recovery.
//!
//! This test cannot run in the Zig test framework because the framework
//! has its own signal handlers. Run via CI script:
//!
//!   ./scripts/check_signal_tests.sh
//!
//! Or manually:
//!   zig build-exe tests/helpers/signal/signal_guard_test.zig src/capi/signal_guard.c -lc
//!   ./signal_guard_test
//!
//! Expected output:
//!   Test 1: Normal operation... PASS
//!   Test 2: SIGBUS recovery... PASS
//!   Test 3: Multiple cycles... PASS
//!   Test 4: Return value propagation... PASS
//!   All tests passed!

const std = @import("std");
const builtin = @import("builtin");

// Import the signal guard C functions (callback-based API)
extern fn talu_signal_guard_install() callconv(.c) void;
extern fn talu_signal_guard_call(
    func: *const fn (?*anyopaque) callconv(.c) c_int,
    ctx: ?*anyopaque,
) callconv(.c) c_int;

const c = @cImport({
    @cInclude("signal.h");
});

// Simple print helper using debug.print (works without buffer setup)
fn print(comptime fmt: []const u8, args: anytype) void {
    std.debug.print(fmt, args);
}

// Test callbacks
fn normalCallback(_: ?*anyopaque) callconv(.c) c_int {
    return 0;
}

fn sigbusCallback(_: ?*anyopaque) callconv(.c) c_int {
    _ = c.raise(c.SIGBUS);
    return 0; // Should not reach here
}

fn returns42(_: ?*anyopaque) callconv(.c) c_int {
    return 42;
}

pub fn main() void {
    if (comptime builtin.os.tag != .linux and builtin.os.tag != .macos) {
        print("Signal guard not supported on this platform\n", .{});
        return;
    }

    talu_signal_guard_install();

    // Test 1: Normal operation (no signal)
    print("Test 1: Normal operation... ", .{});
    {
        const result = talu_signal_guard_call(normalCallback, null);
        if (result == 0) {
            print("PASS\n", .{});
        } else {
            print("FAIL (expected 0, got {d})\n", .{result});
            std.process.exit(1);
        }
    }

    // Test 2: SIGBUS recovery
    print("Test 2: SIGBUS recovery... ", .{});
    {
        const result = talu_signal_guard_call(sigbusCallback, null);
        if (result == -1) {
            print("PASS\n", .{});
        } else {
            print("FAIL (expected -1, got {d})\n", .{result});
            std.process.exit(1);
        }
    }

    // Test 3: Multiple cycles
    print("Test 3: Multiple cycles... ", .{});
    {
        // Cycle 1 - normal
        var result = talu_signal_guard_call(normalCallback, null);
        if (result != 0) {
            print("FAIL (cycle 1: expected 0, got {d})\n", .{result});
            std.process.exit(1);
        }

        // Cycle 2 - with signal
        result = talu_signal_guard_call(sigbusCallback, null);
        if (result != -1) {
            print("FAIL (cycle 2: expected -1, got {d})\n", .{result});
            std.process.exit(1);
        }

        // Cycle 3 - normal again
        result = talu_signal_guard_call(normalCallback, null);
        if (result != 0) {
            print("FAIL (cycle 3: expected 0, got {d})\n", .{result});
            std.process.exit(1);
        }

        print("PASS\n", .{});
    }

    // Test 4: Return value propagation
    print("Test 4: Return value propagation... ", .{});
    {
        const result = talu_signal_guard_call(returns42, null);
        if (result == 42) {
            print("PASS\n", .{});
        } else {
            print("FAIL (expected 42, got {d})\n", .{result});
            std.process.exit(1);
        }
    }

    print("\nAll tests passed!\n", .{});
}
